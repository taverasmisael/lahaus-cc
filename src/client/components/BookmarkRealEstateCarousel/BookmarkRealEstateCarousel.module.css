.carousel {
  @apply w-full relative overflow-hidden rounded-lg hover:opacity-90;
  height: 196px;
}

.overlay {
  @apply w-full h-full absolute top-0 right-0 flex items-center justify-center bg-primary bg-opacity-70;
}
.overlayText {
  @apply text-white text-center text-xl font-bold flex-1;
  margin-left: calc(var(--width) / 2);
}

.image {
  @apply object-cover rounded-lg h-full w-full absolute left-0 top-0 overflow-hidden;
  border: 1px solid transparent;
  max-width: 50%;
}

.image0,
.image1 {
  @apply border-white;
}

.image1,
.image2 {
  --width: calc(calc(60% - 2px) / 2);
  width: var(--width, 33%);
}

.image0 {
  @apply z-20;
}

.image1 {
  @apply z-10;
  left: 45%;
}
.image2 {
  left: calc(40% + var(--width));
}

/* Special styles for the carousel */
/*
* I could have a bit of JS here to make the images sizes and positions a bit more dynamic,
* but I think it's better to keep it simple.
*/

.carousel.singleChild .image {
  @apply max-w-full w-full;
}
.carousel.twoChildren .image {
  @apply w-1/2;
  max-width: 50%;
}
.carousel.twoChildren .image.image0 {
  @apply w-7/12;
  max-width: 60%;
}
.carousel.twoChildren .image.image1 {
  @apply left-1/2;
}
