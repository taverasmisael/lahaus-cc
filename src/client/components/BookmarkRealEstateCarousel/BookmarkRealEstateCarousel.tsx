// next/image was crashing the app on resize for some reason ¯\_(ツ)_/¯
/* eslint-disable @next/next/no-img-element */
import { useMemo } from 'react'
import classNames from 'classnames'
import { RealEstate } from '@/shared/real-state.interface'

import styles from './BookmarkRealEstateCarousel.module.css'

const MAX_IMAGES = 3
const EMPTY_STATE_URL = '/imgs/empty-state.png'

function BookmarkRealEstateCarousel({
  items,
}: BookmarkRealEstateCarouselProps) {
  const images = useMemo(() => getCarouselImages(items), [items])
  const showOverlay = useMemo(() => items.length > MAX_IMAGES, [items])

  return (
    <div
      className={classNames(styles.carousel, {
        [styles.singleChild]: images.length === 1,
        [styles.twoChildren]: images.length === 2,
      })}
    >
      {images.map((image, idx) => (
        <div
          key={idx}
          className={classNames(styles.image, styles[`image${idx}`])}
        >
          <img
            src={image}
            alt={items[idx]?.attributes?.name}
            className="w-full h-full object-cover"
          />
          {idx === 2 && showOverlay && (
            <div data-testid="overlay" className={styles.overlay}>
              <div className={styles.overlayText}>
                +{items.length - MAX_IMAGES}
              </div>
            </div>
          )}
        </div>
      ))}
    </div>
  )
}

interface BookmarkRealEstateCarouselProps {
  items: RealEstate[]
}

export default BookmarkRealEstateCarousel

function getCarouselImages(items: RealEstate[]) {
  const images = items.slice(0, MAX_IMAGES).reduce((prev, item) => {
    const firstImage = item.attributes.gallery_urls[0]
    return firstImage ? [...prev, firstImage] : prev
  }, [] as string[])

  return images.length ? images : [EMPTY_STATE_URL]
}
