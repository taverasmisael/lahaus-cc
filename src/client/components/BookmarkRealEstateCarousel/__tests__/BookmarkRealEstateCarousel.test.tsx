import { render } from '@testing-library/react'

import { RealEstate } from '@/shared/real-state.interface'
import BookmarkRealEstateCarousel from '../BookmarkRealEstateCarousel'

const realEstateFactory = (n = 1) =>
  Array(n)
    .fill(n)
    .map(
      (n) =>
        ({
          id: `${n + 2}`,
          attributes: {
            name: `Real name ${n}`,
            gallery_urls: Array(n)
              .fill(n)
              .map((n) => `https://unsplash.it/${n}`),
          },
        } as RealEstate)
    )

describe('components/BookmarkRealEstateCarousel', () => {
  test('renders empty without crashing', () => {
    const { container } = render(<BookmarkRealEstateCarousel items={[]} />)
    expect(container.firstChild).toMatchSnapshot()
  })
  test('renders single without crashing', () => {
    const { container } = render(
      <BookmarkRealEstateCarousel items={realEstateFactory()} />
    )
    expect(container.firstChild).toMatchSnapshot()
  })
  test('renders double without crashing', () => {
    const { container } = render(
      <BookmarkRealEstateCarousel items={realEstateFactory(2)} />
    )
    expect(container.firstChild).toMatchSnapshot()
  })
  test('renders triple without crashing', () => {
    const { container } = render(
      <BookmarkRealEstateCarousel items={realEstateFactory(3)} />
    )
    expect(container.firstChild).toMatchSnapshot()
  })

  test('renders N without crashing', () => {
    const { container } = render(
      <BookmarkRealEstateCarousel items={realEstateFactory(18)} />
    )
    expect(container.firstChild).toMatchSnapshot()
  })

  test('only displays overlay on 3+ items', () => {
    const { rerender, queryByTestId } = render(
      <BookmarkRealEstateCarousel items={realEstateFactory(3)} />
    )

    expect(queryByTestId('overlay')).toBe(null)

    rerender(<BookmarkRealEstateCarousel items={realEstateFactory(4)} />)

    expect(queryByTestId('overlay')).toHaveTextContent('+1')
  })
})
