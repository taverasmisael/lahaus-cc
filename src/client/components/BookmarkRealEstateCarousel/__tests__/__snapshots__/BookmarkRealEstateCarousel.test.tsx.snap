// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/BookmarkRealEstateCarousel renders N without crashing 1`] = `
<div
  class="carousel"
>
  <div
    class="image image0"
  >
    <img
      alt="Real name 18"
      class="w-full h-full object-cover"
      src="https://unsplash.it/18"
    />
  </div>
  <div
    class="image image1"
  >
    <img
      alt="Real name 18"
      class="w-full h-full object-cover"
      src="https://unsplash.it/18"
    />
  </div>
  <div
    class="image image2"
  >
    <img
      alt="Real name 18"
      class="w-full h-full object-cover"
      src="https://unsplash.it/18"
    />
    <div
      class="overlay"
      data-testid="overlay"
    >
      <div
        class="overlayText"
      >
        +
        15
      </div>
    </div>
  </div>
</div>
`;

exports[`components/BookmarkRealEstateCarousel renders double without crashing 1`] = `
<div
  class="carousel twoChildren"
>
  <div
    class="image image0"
  >
    <img
      alt="Real name 2"
      class="w-full h-full object-cover"
      src="https://unsplash.it/2"
    />
  </div>
  <div
    class="image image1"
  >
    <img
      alt="Real name 2"
      class="w-full h-full object-cover"
      src="https://unsplash.it/2"
    />
  </div>
</div>
`;

exports[`components/BookmarkRealEstateCarousel renders empty without crashing 1`] = `
<div
  class="carousel singleChild"
>
  <div
    class="image image0"
  >
    <img
      class="w-full h-full object-cover"
      src="/imgs/empty-state.png"
    />
  </div>
</div>
`;

exports[`components/BookmarkRealEstateCarousel renders single without crashing 1`] = `
<div
  class="carousel singleChild"
>
  <div
    class="image image0"
  >
    <img
      alt="Real name 1"
      class="w-full h-full object-cover"
      src="https://unsplash.it/1"
    />
  </div>
</div>
`;

exports[`components/BookmarkRealEstateCarousel renders triple without crashing 1`] = `
<div
  class="carousel"
>
  <div
    class="image image0"
  >
    <img
      alt="Real name 3"
      class="w-full h-full object-cover"
      src="https://unsplash.it/3"
    />
  </div>
  <div
    class="image image1"
  >
    <img
      alt="Real name 3"
      class="w-full h-full object-cover"
      src="https://unsplash.it/3"
    />
  </div>
  <div
    class="image image2"
  >
    <img
      alt="Real name 3"
      class="w-full h-full object-cover"
      src="https://unsplash.it/3"
    />
  </div>
</div>
`;
